Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/IniEditorDialog.java

```diff
@@ -0,0 +1,146 @@
+    {
+        final GameFile gameFile =
+                GameFileCacheManager.addOrGet(getArguments().getString(ARG_GAME_PATH));
+        final String gameId = gameFile.getGameId();
+        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+        ViewGroup contents = (ViewGroup) getActivity().getLayoutInflater()
+                .inflate(R.layout.dialog_ini_editor, null);
+
+        TextView textGameId = contents.findViewById(R.id.game_id);
+        textGameId.setText(gameId);
+
+        EditText editCodeContent = contents.findViewById(R.id.code_content);
+        editCodeContent.setHorizontallyScrolling(true);
+        setGameSettings(gameId, editCodeContent);
+
+        Button buttonSave = contents.findViewById(R.id.button_save_ini_file);
+        buttonSave.setOnClickListener(view ->
+        {
+            String content = editCodeContent.getText().toString();
+            SaveIniContent(gameId, content);
+            this.dismiss();
+        });
+
+        builder.setView(contents);
+        return builder.create();
+    }
+
+    private void setGameSettings(String gameId, EditText editCode)
+    {
+        String filename = DirectoryInitialization.getLocalSettingFile(gameId);
+        int count = 0;
+        String[] targets = {"[ActionReplay]", "[ActionReplay_Enabled]", "[Gecko]", "[Gecko_Enabled]"}; // Cheat sections
+        int[] indices = {-1, -1, -1, -1};
+
+        StringBuilder sb = new StringBuilder();
+        try
+        {
+            BufferedReader reader = new BufferedReader(new FileReader(filename));
+            String line = reader.readLine();
+            while (line != null)
+            {
+                for(int i = 0; i < targets.length; ++i)
+                {
+                    int index = line.indexOf(targets[i]);
+                    if(index != -1)
+                    {
+                        indices[i] = count + index;
+                    }
+                }
+                //
+                count += line.length() + 1;
+                sb.append(line);
+                sb.append(System.lineSeparator());
+                line = reader.readLine();
+            }
+        }
+        catch (Exception e)
+        {
+            // ignore
+        }
+
+        int cursorPos = 0;
+        for(int i = 0; i < targets.length; ++i)
+        {
+            if(indices[i] < 0)
+            {
+                sb.append(System.lineSeparator());
+                sb.append(targets[i]);
+                sb.append(System.lineSeparator());
+                count += targets[i].length() + 2;
+                cursorPos = count;
+            }
+            else if(indices[i] > cursorPos)
+            {
+                cursorPos = indices[i] + targets[i].length() + 1;
+            }
+        }
+
+        String content = sb.toString();
+        if(cursorPos > content.length())
+            cursorPos = content.length();
+        editCode.setText(content);
+        editCode.setSelection(cursorPos);
```

Comment:
The logic in this function is needlessly complicated. If you need to do this, then please use the IniFile class – it handles all the string manipulation for you.

Also, I'm not sure if what this function does actually is useful. Perhaps if you're on an outdated fork that doesn't have our cheat UI, always adding the AR and Gecko sections and no other sections is useful, but we're not. If you are going to add any sections by default for the user to use as a template, I think you should do what DolphinWX did: Have a template INI file that you create a copy of when the user tries to access an INI file that doesn't already exist. That would greatly simplify the code.